engine/src/main/java/org/terasology/rendering/dag/RenderGraph.java

```diff
@@ -15,44 +15,138 @@
 
-    public RenderGraph() {
-        nodes = Lists.newArrayList();
-    }
+    private Map<SimpleUri, Node> nodeMap = Maps.newHashMap();
+    private Multimap<Node, Node> edgeMap = HashMultimap.create();
+    private Multimap<Node, Node> reverseEdgeMap = HashMultimap.create();
 
     public SimpleUri addNode(Node node, String suggestedUri) {
-        nodes.add(node);
+        Validate.notNull(node, "node cannot be null!");
+        Validate.notNull(suggestedUri, "suggestedUri cannot be null!");
 
-        // TODO: make sure URIs are actually unique: if "myModule:blur" is present the node gets the uri "myModule:blur2" instead.
-        // TODO: make sure the namespace in the uri is engine-assigned, so that only engine nodes can have the "engine:" namespace - everything else gets the namespace of the module.
         SimpleUri nodeUri = new SimpleUri("engine:" + suggestedUri);
+
+        if (nodeMap.containsKey(nodeUri)) {
+            int i = 2;
+            while (nodeMap.containsKey(new SimpleUri(nodeUri.toString() + i))) {
+                i++;
+            }
+            nodeUri = new SimpleUri(nodeUri.toString() + i);
+        }
+
+        nodeMap.put(nodeUri, node);
         node.setUri(nodeUri);
+
         return nodeUri;
     }
 
+    public Node removeNode(SimpleUri nodeUri) {
+        Validate.notNull(nodeUri, "nodeUri cannot be null!");
+
+        if (edgeMap.containsKey(nodeUri)) {
+            throw new RuntimeException("The node you are trying to remove is still connected to other nodes in the graph!");
+        }
+
+        return nodeMap.remove(nodeUri);
+    }
+
     public Node findNode(SimpleUri nodeUri) {
-        for (Node node: nodes) {
-            if (node.getUri().equals(nodeUri)) {
-                return node;
-            }
+        Validate.notNull(nodeUri, "nodeUri cannot be null!");
+
+        return nodeMap.get(nodeUri);
+    }
+
+    public boolean connect(Node fromNode, Node toNode) {
+        Validate.notNull(fromNode, "fromNode cannot be null!");
+        Validate.notNull(toNode, "toNode cannot be null!");
+
+        reverseEdgeMap.put(toNode, fromNode);
+        boolean success = edgeMap.put(fromNode, toNode);
+        if (!success) {
+            logger.warn("Trying to connect two already connected nodes, " + fromNode.getClass() + " and " + toNode.getClass());
         }
 
-        return null;
+        return success;
+    }
+
+    public boolean disconnect(Node fromNode, Node toNode) {
+        Validate.notNull(fromNode, "fromNode cannot be null!");
+        Validate.notNull(toNode, "toNode cannot be null!");
+
+        reverseEdgeMap.remove(toNode, fromNode);
+        return edgeMap.remove(fromNode, toNode);
     }
 
-    // TODO: add remove, get, addEdge, removeEdge methods here
+    // TODO: Add `boolean isFullyFunctional(Node node)`
+
+    // TODO: Add handler methods which the graph uses to communicate changes to a node.
 
     public List<Node> getNodesInTopologicalOrder() {
```

Comment:
From what I know, I have implemented Kahn's Algorithm.

Here is the concrete reference I used: https://www.geeksforgeeks.org/topological-sorting-indegree-based-solution/

Note that the wiki uses steps like `if m has no other incoming edges then` and `remove edge e from the graph`. Since (cloning and) modifying the actual graph is not a great idea, we simply create a new map (the inDegreeMap) that keeps track of the incomingDegree of each node (since we don't really care about outgoing edges, except when removing edges). This helps convert the abstract steps of the wiki into concrete steps of the algorithm. At its core, my algorithm is the same.

Reply:
I guess I would have preferred the wikipedia version because it's likely to be more stable from a documentation point of view. geeksforgeeks.org could be shut down tomorrow for what I know.

The other thing I don't like is that apart from the title the geeksforgeeks.org page provides no reference/credit to Kahn's original paper.

Nevertheless: let's keep the code as is but let's provide (as in-code comments) a link to the g4g.org page and a mention to Kahn (1962), which is the original paper.